### PR TITLE
Ligatures checked against PragmataPro, Hasklig and Fira Code

### LINKED ISSUE #49: Ligature support is being removed in some cases.
Atom v1.1.0 now officially supports ligatures. When this package is active, not all ligatures are displaying. 

An example: The left frame has a Julia file and the right an untitled document. The untitled document has `!=` and `<=` ligatures working, although once saved into a .lj and syntax highlighted, the ligatures are no longer visible.
julia

Not all ligatures are removed, for example `->` still works. The problem doesn't occur in other file types (for example a .cpp or .hs file, which aren't syntax highlighted via this package), and I've tested this with most of the ligature enabled fonts: no changes there either.

### PR DESCRIPTION
**Pull request by Libbum**
Fixes #49, although it'd be good if we could get a few people to test before we merge. I've checked against the current ligature sets that are legal Julia and reside within the three fonts that have been discussed so far: Pragmata Pro, Hasklig and Fira Code. 

I've added a few missing operators and fixed a couple that weren't matching properly. To get some of the ligatures working I've had to add in a few negative look aheads and two negative look behinds to keep everything sane, otherwise we had problems differentiating between `<`, `<<`, `>>=`, `>>>`, `>>>=` for example. I've also added a new category for the function applicator `|>` called `applies`. It was one of the missing operators I've added, but didn't think it fit into the other categories. 

I think I've covered all future ligature additions with this as they come up, but can't be certain. Pragmata Pro fails gracefully when it can't complete the whole operator match (for example the dot operators like `.<=`, will print a two-part operator using `.` and a `<=` ligature). The other two fonts have a match each that doesn't look right at the moment: Hasklig matches `/`,`/=` in the `//=` operators; Fira Code matches `.=`,`=` in the `.==` operator.  If we want to futureproof, there's not much can be done about this for now.

### CODE REVIEW

**Comment by spencerlyon2** [comment]
Looks good to me. Should I merge, or wait for feedback from others first? What do you think @Libbum ?

**Comment by Libbum** [comment]
There's zero rush from me. So if you want to maybe wait a week and see if anyone else has input that's totally fine. But I'll leave it to your discretion.

File: grammars/julia.cson [hunk 1]
```diff
@@ -218,11 +218,15 @@ repository:
...
+        match: "(?:<<(?!=)|>>>(?!=)|>>(?!=|>=)|\\.>>>|\\.>>|\\.<<)"
+        name: "keyword.operator.shift.julia"
+      }
+      {
+        match: "(?:===|\\.==|!==|!=|\\.>=|\\.>(?!=)|\\.<(?!=)|==|\\.!=|\\.=|\\.!|<:|:>|(?<!>)>=|(?<!<)<=|\\.<=|>(?!>)|<(?!<))"
         name: "keyword.operator.relation.julia"
       }
       {
-        match: "(?::=|\\+=|-=|\\*=|//=|/=|\\.//=|\\.\\*=|\\\\=|\\.\\\\=|^=|\\.^=|%=|\\|=|&=|\\$=|<<=|>>=|=)"
+        match: "(?::=|\\+=|-=|\\*=|//=|/=|\\.//=|\\./=|\\.\\*=|\\\\=|\\.\\\\=|\\^=|\\.\\^=|%=|\\.%=|÷=|\\.÷=|\\|=|&=|\\$=|<<=|>>=|>>>=|=)"
```
Comment (50Wliu) [inline grammars/julia.cson]: Can't these be moved above `keyword.operator.shift` to get rid of those negative lookaheads?

File: grammars/julia.cson [hunk 2]
```diff
@@ -218,11 +218,15 @@ repository:
         name: "keyword.operator.arrow.julia"
       }
       {
-        match: "(?:\\.==|!=|\\.>=|\\.>=|\\.>|\\.<|==|\\.!=|\\.=|\\.!|<:|:>|>=|<=|>|<)"
+        match: "(?:<<(?!=)|>>>(?!=)|>>(?!=|>=)|\\.>>>|\\.>>|\\.<<)"
+        name: "keyword.operator.shift.julia"
+      }
+      {
+        match: "(?:===|\\.==|!==|!=|\\.>=|\\.>(?!=)|\\.<(?!=)|==|\\.!=|\\.=|\\.!|<:|:>|(?<!>)>=|(?<!<)<=|\\.<=|>(?!>)|<(?!<))"
```
Comment (50Wliu) [inline grammars/julia.cson]: I'm pretty sure the lookaheads for `\\.>(?!=)` and `\\.<(?!=)` shouldn't be necessary in the first place - were ligatures not working for those beforehand?  The others are unnecessary if `keyword.operator.update` is placed above these.

File: grammars/julia.cson [hunk 3]
```diff
@@ -246,11 +246,15 @@ repository:
...
+        match: "(?:\\|>)"
+        name: "keyword.operator.applies.julia"
+      }
+      {
         match: "(?:\\||\\&|~)"
         name: "keyword.operator.bitwise.julia"
       }
       {
-        match: "(?:\\+|-|\\*|\\.\\*|/|\\./|//|\\.//|%|\\.%|\\\\|\\.\\\\|\\^|\\.\\^)"
+        match: "(?:\\+\\+|--|\\+|\\.\\+|-|\\.\\-|\\*|\\.\\*|//(?!=)|\\.//(?!=)|/|\\./|%|\\.%|\\\\|\\.\\\\|\\^|\\.\\^)"
```
Comment (50Wliu) [inline grammars/julia.cson]: Again, are these lookaheads necessary?  The ones in `keyword.operator.update` should definitely be getting precedence over these.

**Comment by spencerlyon2** [comment]
@Libbum do you mind adding some tests for the new regexes you implemented for better operator scoping?

**Comment by Libbum** [comment]
@50Wliu most of those suggestions worked out well. The only reason I started there was to keep update on the bottom so that `=` wasn't matched all the time. So with your suggestion we can remove 5 negative look aheads for the cost of adding one onto `=` (to catch `==` and `===` in relation).

Sorry the diff looks like a mess, I really didn't change much other than dropping the negative look aheads and swapping the order. 

@spencerlyon2 Sure thing. Will get back to you.

**Comment by Libbum** [comment]
OK, I've never built a test like this. Seems I've made it fail. @spencerlyon2: can you take a quick look, is there something I've missed here?

Getting me to check was a good idea though: I noticed the ternary operator (`?:`) wasn't parsed right, so I've fixed that at least.

**Comment by spencerlyon2** [comment]
It looks like the tests that are now failing are the ones that try to test if we can tokenize symbols. 

Specifically, the changes here seem to have changed things like this

```julia
:my_symbol
```

before it would say that `:` is `keyword.operator.ternary.julia`

Not it says that the `:` is a `keyword.operator.range.julia`.

I don't know that either of them are accurate, but fixing that might be out of the scope of this PR. What do you think?

**Comment by Libbum** [comment]
Oh yes, now I see. 
Yeah, I agree neither of these are necessarily 'correct'. It looks like the choice of matching ternary there is essentially because a more complete symbol matching logic is going to be a little more difficult to implement. 

So perhaps the plan here would be to do the following:
1. Change the symbol check match on line 126 from `keyword.operator.ternary.julia` to `keyword.operator.range.julia`
2. Open a new issue to investigate the correct scoping of symbols

Sound good?

**Comment by spencerlyon2** [comment]
That sounds good to me.

Note that you will have to change that scoping rule on all the filled tests, not just that one.

Thanks again for working on this

**Comment by Libbum** [comment]
Travis-ci had some issues booting OS X jobs (and I put in a few extra spaces in the wrong places), but everything seems to build now.

I'll write up a new symbol scoping issue tomorrow when I have the time.

**Comment by spencerlyon2** [comment]
Thanks.

**Pull request merged**